### DISCUSSION TITLE
Add related objects descriptors to related objects on subclasses?

### DISCUSSION THREAD

**Discussion by blag** [jazzband/django-polymorphic#718] [Q&A]
When subclassing from a PolymorphicModel with a relation to another model, it would be awesome if the related model had reverse related descriptors added to it for all subclasses of the polymorphic model.

This is probably better illustrated with an example:

``` python
class Company(models.Model):
    ...

class DeviceModel(PolymorphicModel):
    manufacturer = models.ForeignKey(Company, related_name='device_models')
    ...

class TabletModel(DeviceModel):
    ...

class PhoneModel(DeviceModel):
    ...
```

All `Company` objects have a reverse related descriptor to the polymorphic superclass:

``` python
apple = Company.objects.get(name='Apple')
apple.device_models.all()  # Returns all device models
```

but they don't have reverse related descriptors to the subclasses:

``` python
apple.tabletmodels_set.all()
apple.phonemodels_set.all()
```

It would be super cool if the related class (`Company`) had more reverse related descriptors to access subsets of the polymorphic model, like so:

``` python
class Company(models.Model):
    ...

class DeviceModel(PolymorphicModel):
    manufacturer = models.ForeignKey(Company, related_name='%(class)ss')
    ...

class TabletModel(DeviceModel):
    ...

class PhoneModel(DeviceModel):
    ...

apple = Company.objects.get(name='Apple')
apple.devicemodels.all()  # All device models, including objects of subclasses
apple.tabletmodels.all()  # Only device models that are tablets, using the `TabletModel`'s default manager
apple.phonemodels.all()  # Only device models that are phones, using the `PhoneModel`'s default manager
```

This is would make subclassing a polymorphic model a lot more like subclassing an abstract model from vanilla Django.

If there is a known workaround for this I haven't been able to figure it out. Issue #19 looked close but I didn't get it to work since I need `DeviceModel` to be a concrete class and I didn't want to have foreign keys to `app_company` table on all of the `app_devicemodel`, `app_tabletmodel`, and `app_phonemodel` tables. Any help is greatly appreciated.

**Comment by bckohan** [comment]
@blag You could do this using the [instance_of](https://django-polymorphic.readthedocs.io/en/stable/api/polymorphic.query.html#polymorphic.query.PolymorphicQuerySet.instance_of) filter on the device models related manager:

```python
apple.device_models.instance_of(PhoneModel)  # only phone models
```

You might also want to add some convenience functions to your default DeviceModel manager:

```python
from polymorphic.query import PolymorphicQuerySet

class DeviceQuerySet(PolymorphicQuerySet)
    def phones(self):
        return self.instance_of(PhoneModel)
    def tablets(self):
        return self.instance_of(TabletModel)

class DeviceModel(PolymorphicModel):
    objects = DeviceQuerySet.as_manager()
    manufacturer = models.ForeignKey(Company, related_name='devices')

apple = Company.objects.get(name='Apple')
apple.devices.all()  # All device models, including objects of subclasses
apple.devices.phones()  # Only device models that are tablets, using the `TabletModel`'s default manager
apple.devices.tablets()  # Only device models that are phones, using the `PhoneModel`'s default manager
```
